src/Strategy/Node/Pnpm/PnpmLock.hs

```diff
@@ -0,0 +1,298 @@
+      <|> (RegistryResolution <$> (RegistryPnpmResolution <$> obj .: "integrity"))
+
+analyze :: (Has ReadFS sig m, Has Diagnostics sig m) => Path Abs File -> m (Graphing Dependency)
+analyze file = context "Analyzing Npm Lockfile (v3)" $ do
+  pnpmLockFile <- context "Parsing pnpm-lock file" $ readContentsYaml file
+  context "Building dependency graph" $ pure $ buildGraph pnpmLockFile
+
+buildGraph :: PnpmLockfile -> Graphing Dependency
+buildGraph lockFile = withoutLocalPackages $
+  run . evalGrapher $ do
+    for_ (toList $ importers lockFile) $ \(_, projectSnapshot) -> do
+      let allDirectDependencies =
+            toList (directDependencies projectSnapshot)
+              <> toList (directDevDependencies projectSnapshot)
+
+      for_ allDirectDependencies $ \(depName, depVersion) -> do
+        case (toResolvedDependency depName depVersion) of
+          Just dep -> direct dep
+          Nothing -> pure ()
+
+    -- Add edges and deep dependencies by iterating over all packages.
+    --
+    -- Use `dev` to infer if this is production or non-production dependency.
+    -- Use `dependencies` to infer the edge from this dependency (aws-sdk@2.1148.0)
+    --   to its children. If the child entry cannot be found in `packages`,
+    --   do not provision an edge.
+    --
+    -- @
+    -- > packages:
+    -- >
+    -- >  /aws-sdk/2.1148.0:
+    -- >    resolution: {integrity: sha512-FUYAyveKmS5eq..==}
+    -- >    engines: {node: '>= 10.0.0'}
+    -- >    dependencies:
+    -- >      buffer: 4.9.2
+    -- >      events: 1.1.1
+    -- >    dev: false
+    -- @
+    for_ (toList $ packages lockFile) $ \(pkgKey, pkgMeta) -> do
+      let deepDependencies =
+            toList (dependencies pkgMeta)
+              <> toList (peerDependencies pkgMeta)
+
+      let (depName, depVersion) = case getPkgNameVersion pkgKey of
+            Nothing -> (pkgKey, Nothing)
+            Just (name, version) -> (name, Just version)
+      let parentDep = toDependency depName depVersion pkgMeta
+
+      -- It is ok, if this dependency was already graphed as direct
+      -- @direct 1 <> deep 1 = direct 1@
+      deep parentDep
+
+      for_ deepDependencies $ \(deepName, deepVersion) -> do
+        case (toResolvedDependency deepName deepVersion) of
+          Just childDep -> edge parentDep childDep
+          Nothing -> pure ()
+  where
+    -- Gets package name and version from package's key.
+    --
+    -- >> getPkgNameVersion "" = Nothing
+    -- >> getPkgNameVersion "github.com/something" = Nothing
+    -- >> getPkgNameVersion "/pkg-a/1.0.0" = Just ("pkg-a", "1.0.0")
+    getPkgNameVersion :: Text -> Maybe (Text, Text)
+    getPkgNameVersion pkgKey = case (Text.stripPrefix "/" pkgKey) of
+      Nothing -> Nothing
+      Just txt -> do
+        let (nameWithSlash, version) = Text.breakOnEnd "/" txt
+        case (Text.stripSuffix "/" nameWithSlash, version) of
+          (Just name, v) -> Just (name, v)
+          _ -> Nothing
```

Comment:
[nit]: I think a lot of this pattern matching can be done with `>>=` instead.  Might be worth refactoring.